api/src/main/java/com/epam/pipeline/manager/datastorage/providers/gcp/GSBucketStorageHelper.java

```diff
@@ -378,6 +381,72 @@ public void restoreFileVersion(final GSBucketStorage storage, final String path,
         deleteBlob(blob, client, true);
     }
 
+    public void restoreFolder(final GSBucketStorage storage, final String path,
+                              final RestoreFolderVO restoreFolderVO) {
+        final Storage client = gcpClient.buildStorageClient(region);
+        final String bucketName = storage.getPath();
+        cleanDeleteMarkers(client, bucketName, path, restoreFolderVO);
+    }
+
+    private void cleanDeleteMarkers(final Storage client,
+                                    final String bucketName, final String requestPath,
+                                    final RestoreFolderVO restoreFolderVO) {
+        String folderPath = Optional.ofNullable(requestPath).orElse(EMPTY_PREFIX);
+        if (StringUtils.isNotBlank(folderPath)) {
+            folderPath = normalizeFolderPath(requestPath);
+        }
+        final Page<Blob> blobs = client.list(bucketName,
+                Storage.BlobListOption.versions(true),
+                Storage.BlobListOption.currentDirectory(),
+                Storage.BlobListOption.prefix(folderPath),
+                Storage.BlobListOption.pageToken(EMPTY_PREFIX),
+                Storage.BlobListOption.pageSize(Integer.MAX_VALUE));
+        Assert.isTrue(Objects.nonNull(blobs) && blobs.iterateAll().iterator().hasNext(), messageHelper
+                .getMessage(MessageConstants.ERROR_DATASTORAGE_PATH_NOT_FOUND, folderPath, bucketName));
+        listItemsWithVersions(blobs)
+                .forEach(item -> {
+                    recursiveRestoreFolderCall(item, client, bucketName, restoreFolderVO);
+                    if (isFileWithDeleteMarkerAndShouldBeRestore(item, restoreFolderVO)) {
+                        final Blob blob = checkBlobExistsAndGet(bucketName, item.getPath(), client,
+                                removeDeletedMarkerFromVersion(((DataStorageFile) item).getVersion()));
+                        final Storage.CopyRequest request = Storage.CopyRequest.newBuilder()
```

Comment:
Maybe it would be better to create a new method (wth copy bob logic) and reuse it into `restoreFileVersion`?

Reply:
done